### ISSUE TITLE
ServicePartitionResolver hangs when tries to resolve service in error state

### ISSUE THREAD

**Issue by OlegKarasik** [Azure/service-fabric-issues#918]
The following code with hang if ```fabric:/MyApp/MySrv``` is currently in error state (In my case service was existing all the time because of exception).

The code was executed from ASP.NET Core middleware handler.

``` C#

var resolver = ServicePartitionResolver.GetDefault();
var partition = await resolver.ResolveAsync(new Uri("fabric:/MyApp/MySrv"), new ServicePartitionKey(), CancellationToken.None);

```

**Comment by vturecek** [comment]
@oanapl I wouldn't expect name resolution to have anything to do with the health state of the target service..

**Comment by oanapl** [comment]
No, there is no direct correlation between health state and resolution. But the fact that the service keeps crashing may have something to do with the failure :) 
@BharatNarasimman , can you take a look?

**Comment by OlegKarasik** [comment]
@BharatNarasimman

Probably this could be important. The service ``` fabric:/MyApp/MySrv ``` was never started successfully.

The process was something like:
1. Deploy service ``` fabric:/MyApp/MySrv ``` (the deployment ended - type was registered etc.) but the service was all the time failing.
2. Deploy service with the above code.

**Comment by suchiagicha** [comment]
@OlegKarasik 
Resolve method retries on all transient exceptions. For cases where you want to limit the max execution time of this method, you should create a cancelation token.
https://msdn.microsoft.com/en-us/library/system.threading.cancellationtoken(v=vs.110).aspx

**Comment by OlegKarasik** [comment]
@suchiagicha 
OK. I see this works when you have meaningful cancellation token (i.e. you inherit it from the HttpContext or some) but does this make sense to have some default ```maxRetries``` or ```timeout``` reaching what ```ResolveAsync``` would trigger an exception?

**Comment by suchiagicha** [comment]
@OlegKarasik 
We think cancellation token should be enough for those cases.

**Comment by OlegKarasik** [comment]
@suchiagicha 
I think support for timeout would be great for those cases but at the same time I  agree with you that cancellation token is technically enough (everything else can be implemented on top of this).

Still I am not sure that maintaining cancellation token here is the right solution. In this case the service and its partitions are in the error state and from my prospective attempt to communicate with the broken service differs from the use case when [service / naming service / etc] simply timed out or encountered a network issue.. 

Should the attempt to contact the service / partition that is in error state result in the exception as when you are trying to resolve the service / partition that doesn't exist?

**Comment by suchiagicha** [comment]
@OlegKarasik 
From the client side (ServicePartitionResolver) you cant differentiate if there is a network issue or service is in error state(which am assuming is crashing) . You can use health Api to check the health of the service in order to differentiate the above issue.

**Comment by suchiagicha** [comment]
I got your point. Will improve the exception handling in this case.
Thankyou :)
